File: src/main/java/kitchenpos/menu/tobe/domain/MenuProduct.java
package kitchenpos.menu.tobe.domain;

import jakarta.persistence.*;
import kitchenpos.exception.IllegalQuantityException;

import java.util.UUID;

@Table(name = "menu_product")
@Entity
public class MenuProduct {

    @Column(name = "seq")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private long seq;

    @ManyToOne
    @JoinColumn(name = "menu_id", nullable = false)
    private Menu menu;

    private UUID productId;

    @Column(name = "quantity", nullable = false)
    private long quantity;

    private long price;

    protected MenuProduct() {

    }

    public static MenuProduct of(final UUID productId, final long price,  final long quantity) {
        validateQuantity(quantity);
        return new MenuProduct(productId, price, quantity);
    }

    private MenuProduct(final UUID productId, final long price,  final long quantity) {
        this.productId = productId;
        this.price = price;
        this.quantity = quantity;
    }

    private static void validateQuantity(long quantity) {
        if (quantity < 0) {
            throw new IllegalQuantityException("수량은 0개보다 적을 수 없습니다.");
        }
    }

    public long getMenuProductPrice() {
        return price * quantity;
    }


    public UUID getProductId() {
        return productId;
    }

    public long getQuantity() {
        return quantity;
    }

    public long getPrice() {
        return price;
    }

}

File: src/main/kotlin/kitchenpos/menu/tobe/domain/MenuProduct.kt
package kitchenpos.menu.tobe.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.util.*

@Table(name = "menu_product")
@Entity(name = "TobeMenuProduct")
class MenuProduct(
    @Column(name = "seq")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    var seq: Long? = null,

    @Column(name = "product_id")
    var productId: UUID,

    @Column(name = "quantity", nullable = false)
    val quantity: Long
) {

    init {
        require(quantity >= 0) { "메뉴 상품의 수량은 0개 이상이어야 합니다." }
    }
}
